src/components/appointments/AppointmentFields.js

```diff
@@ -0,0 +1,189 @@
+import React, {useContext, useState} from 'react';
+import {FormattedMessage, useIntl} from 'react-intl';
+import {useAsync} from 'react-use';
+
+import {ConfigContext} from 'Context';
+import {get} from 'api';
+import Label from 'components/Label';
+import Loader from 'components/Loader';
+import {getBEMClassName, getFormattedDateString, getFormattedTimeString} from 'utils';
+
+const requestOrEmpty = async (url, query, condition) => {
+  if (condition) {
+    return await get(url, query);
+  } else {
+    return [];
+  }
+};
+
+const ProductField = ({value, onChange, url}) => {
+  const {loading, value: items} = useAsync(async () => await get(url));
+  // todo process errors
+
+  return (
+    <div className={getBEMClassName('form-control')}>
+      <Label>
+        <FormattedMessage description="Appointment product field label" defaultMessage="Product" />
+      </Label>
+
+      {loading ? (
+        <Loader />
+      ) : (
+        <select className={getBEMClassName('select')} value={value} onChange={onChange}>
+          <option value={''}></option>
+          {items.map(item => (
+            <option value={item.identifier} key={item.identifier}>
+              {item.name}
+            </option>
+          ))}
+        </select>
```

Comment:
there's a lot of low-level HTML repetition for these dropdowns, I'd drop in react-select instead. It can also handle async stuff out of the box.

Reply:
To be honest I didn't managed react-select to work with async requests. But for now all low-level code is encapsulated in `AppointmentField` component and it's more DRYer